components/resource_metering/src/recorder/mod.rs

```diff
@@ -0,0 +1,426 @@
+use tikv_util::time::Instant;
+use tikv_util::worker::Scheduler;
+
+mod cpu;
+
+pub use cpu::CpuRecorder;
+
+const RECORD_FREQUENCY: f64 = 99.0;
+const RECORD_LEN_THRESHOLD: usize = 20_000;
+const CLEANUP_INTERVAL_SECS: u64 = 15 * 60;
+
+/// This trait defines a general framework that works at a certain frequency. Typically,
+/// it describes the recorder(sampler) framework for a specific resource.
+///
+/// [Recorder] will maintain a list of sub-recorders, driving all sub-recorders to work
+/// according to the behavior described in this trait.
+pub trait SubRecorder {
+    /// This function is called at a fixed frequency. (A typical frequency is 99hz.)
+    ///
+    /// The [RawRecords] and [LocalStorage] map of all threads will be passed in through
+    /// parameters. We need to collect resources (may be from each `LocalStorage`) and
+    /// write them into `RawRecords`.
+    ///
+    /// The implementation needs to sample the resource in this function (in general).
+    ///
+    /// [RawRecords]: crate::model::RawRecords
+    /// [LocalStorage]: crate::localstorage::LocalStorage
+    fn tick(
+        &mut self,
+        _records: &mut RawRecords,
+        _thread_stores: &mut HashMap<usize, LocalStorage>,
+    ) {
+    }
+
+    /// This function is called every time before reporting to Collector.
+    /// The default period is 1 second.
+    ///
+    /// The [RawRecords] and [LocalStorage] map of all threads will be passed in through parameters.
+    ///
+    /// [RawRecords]: crate::model::RawRecords
+    /// [LocalStorage]: crate::localstorage::LocalStorage
+    fn collect(
+        &mut self,
+        _records: &mut RawRecords,
+        _thread_stores: &mut HashMap<usize, LocalStorage>,
+    ) {
+    }
+
+    /// This function is called when we need to clean up data.
+    /// The default period is 5 minutes.
+    fn cleanup(&mut self) {}
+
+    /// This function is called when a reset is required.
+    ///
+    /// The typical situation is when the [Recorder] thread is suspended.
+    fn reset(&mut self) {}
+
+    /// This function is called when a new thread accesses thread-local-storage.
+    ///
+    /// This function exists because the sampling work of `SubRecorder` may need
+    /// to be performed on all functions, and `SubRecorder` may wish to maintain
+    /// a thread-related data structure by itself.
+    fn thread_created(&mut self, _id: usize, _tag: SharedTagPtr) {}
+}
+
+/// Give `Recorder` a list of [SubRecorder]s and `Recorder` will make them work
+/// correctly. That's it.
+///
+/// All SubRecorders run on the same recorder thread.
+///
+/// We cannot construct `Recorder` directly, but need to construct it through
+/// [RecorderBuilder]. We can pass the `SubRecorder` (and other parameters)
+/// that the `Recorder` needs to load through the `RecorderBuilder`.
+pub struct Recorder<C> {
+    pause: Arc<AtomicBool>,
+    precision_ms: Arc<AtomicU64>,
+    records: RawRecords,
+    collector: C,
```

Comment:
Why is there a specific collector here? What is the difference between it and collectors inside `dync_collectors`?